### DISCUSSION TITLE
v2-to-v3 transformation discussion

### DISCUSSION THREAD

**Discussion by Zefau** [Zefau/ioBroker.jarvis#769] [General]
This thread is for discussion, questions and issues related to jarvis v3 (and the transformation from v2). Issue are collected in https://github.com/Zefau/ioBroker.jarvis/issues/770

**Comment by sigi2345** [comment]
<details> <summary>Test</summary>


jarvis.0 | 2021-06-05 16:16:15.869 | info | WebSocket opening on port 8400 using https...
-- | -- | -- | --
jarvis.0 | 2021-06-05 16:16:15.821 | error | Cannot configure secure web server, because no certificates found: , ,
jarvis.0 | 2021-06-05 16:16:15.798 | info | Found Backups for devices.
jarvis.0 | 2021-06-05 16:16:15.797 | info | Found Backups for styles.
jarvis.0 | 2021-06-05 16:16:15.796 | info | Found Backups for settings.
jarvis.0 | 2021-06-05 16:16:15.791 | info | Found Backups for layout.
jarvis.0 | 2021-06-05 16:16:15.778 | info | starting. Version 3.0.0-alpha.0 in C:/Program Files/iobroker/Home/node_modules/iobroker.jarvis, node: v12.22.1, js-controller: 3.3.11
jarvis.0 | 2021-06-05 16:16:10.942 | info | Terminated (ADAPTER_REQUESTED_TERMINATION): Without reason
jarvis.0 | 2021-06-05 16:16:10.942 | info | terminating
jarvis.0 | 2021-06-05 16:16:10.941 | info | Adapter stopped und unloaded.
jarvis.0 | 2021-06-05 16:16:10.941 | info | Got terminate signal TERMINATE_YOURSELF
jarvis.0 | 2021-06-05 16:16:10.880 | info | Found Backups for devices.
jarvis.0 | 2021-06-05 16:16:10.869 | info | Found Backups for layout.
jarvis.0 | 2021-06-05 16:16:10.868 | info | Found Backups for styles.
jarvis.0 | 2021-06-05 16:16:10.867 | info | Found Backups for settings.
jarvis.0 | 2021-06-05 16:16:10.863 | info | WebSocket opening on port 8400 using https...
jarvis.0 | 2021-06-05 16:16:10.828 | error | Cannot configure secure web server, because no certificates found: undefined, undefined, undefined
jarvis.0 | 2021-06-05 16:16:10.805 | info | starting. Version 3.0.0-alpha.0 in C:/Program Files/iobroker/Home/node_modules/iobroker.jarvis, node: v12.22.1, js-controller: 3.3.11
jarvis.0 | 2021-06-05 16:14:28.521 | info | Terminated (ADAPTER_REQUESTED_TERMINATION): Without reason
jarvis.0 | 2021-06-05 16:14:28.520 | info | terminating
jarvis.0 | 2021-06-05 16:14:28.520 | info | Adapter stopped und unloaded.
jarvis.0 | 2021-06-05 16:14:28.519 | info | Got terminate signal TERMINATE_YOURSELF



</details>

**Comment by Zefau** [comment]
Fängt ja gut an ;-)  hast du Web Adapter mit https konfiguriert?

**Reply by sigi2345** [reply]
> Fängt ja gut an ;-) hast du Web Adapter mit https konfiguriert?

Nein

**Reply by Zefau** [reply]
Ok, schau ich nachher mal

**Reply by sigi2345** [reply]
> Ok, schau ich nachher mal

Test 2 sieht besser aus, aber noch nicht Richtig.

`



jarvis.0 | 2021-06-05 23:16:10.919 | info | Client with IP ::1 connected
-- | -- | -- | --
jarvis.0 | 2021-06-05 23:14:16.600 | info | Backup devices..
jarvis.0 | 2021-06-05 23:14:16.318 | info | Backup layout..
jarvis.0 | 2021-06-05 23:14:15.777 | warn | State "jarvis.0.widgets" has no existing object, this might lead to an error in future versions
jarvis.0 | 2021-06-05 23:14:15.125 | info | undefined is not a valid state value for id "jarvis.0.clients.1.newNotifications"
jarvis.0 | 2021-06-05 23:14:15.099 | info | Client with IP ::1 connected
jarvis.0 | 2021-06-05 23:13:45.863 | info | WebSocket opening on port 8400 using http...
jarvis.0 | 2021-06-05 23:13:45.824 | info | Found Backups for devices.
jarvis.0 | 2021-06-05 23:13:45.822 | info | Found Backups for styles.
jarvis.0 | 2021-06-05 23:13:45.821 | info | Found Backups for settings.
jarvis.0 | 2021-06-05 23:13:45.817 | info | Found Backups for layout.
jarvis.0 | 2021-06-05 23:13:45.804 | info | starting. Version 1.0.0-alpha.1 in C:/Program Files/iobroker/Home/node_modules/iobroker.jarvis, node: v12.22.1, js-controller: 3.3.11
jarvis.0 | 2021-06-05 23:11:17.893 | info | Terminated (NO_ERROR): Without reason
jarvis.0 | 2021-06-05 23:11:17.892 | info | terminating
jarvis.0 | 2021-06-05 23:11:17.890 | info | Got terminate signal TERMINATE_YOURSELF
jarvis.0 | 2021-06-05 23:11:17.884 | info | Adapter stopped und unloaded.
jarvis.0 | 2021-06-05 23:11:17.881 | info | Adapter is disabled =&gt; stop



`

![Screenshot (3863)](https://user-images.githubusercontent.com/41079106/120905775-af90ff80-c654-11eb-9c33-2ba31859441c.png)
![Screenshot (3864)](https://user-images.githubusercontent.com/41079106/120905776-b0299600-c654-11eb-92a1-6cfee30eac0e.png)

**Reply by Zefau** [reply]
Sieht gut aus. Was hattest du dort für Module, wo es nun leer ist?

**Reply by sigi2345** [reply]
Sehr viele gemischte. Sind leider weg?

**Reply by Zefau** [reply]
Was steht in der Browser Konsole? Wird dort ein Fehler angezeigt?

**Reply by sigi2345** [reply]
> Was steht in der Browser Konsole? Wird dort ein Fehler angezeigt?

![Screenshot (3866)](https://user-images.githubusercontent.com/41079106/120906132-2fb86480-c657-11eb-9d73-43a4c6fa8b19.png)

**Reply by Zefau** [reply]
Jetzt hast du den Info Reiter selektiert. Was steht bei den Warnings?

**Reply by sigi2345** [reply]
![Screenshot (3869)](https://user-images.githubusercontent.com/41079106/120906204-c1c06d00-c657-11eb-86b7-2d08887fc1d4.png)

**Reply by Zefau** [reply]
ah, ich glaube es liegt daran, dass bei dir `jarvis.0.widgets` nicht angelegt wurde?

**Reply by sigi2345** [reply]
Wo , in den Objekten?

![Screenshot (3873)](https://user-images.githubusercontent.com/41079106/120906442-ac4c4280-c659-11eb-85db-240e349039d5.png)

**Reply by Zefau** [reply]
Zieh dir mal die alpha.3 und starte nochmal mit der v2 Konfiguration. jarvis konvertiert beim Start dann in die v3 und befüllt `jarvis.0.widgets`.

**Reply by sigi2345** [reply]
Widgets sind auf 0

**Reply by Zefau** [reply]
Was meinst du mit 0?

**Comment by mcuiobroker** [comment]
Nach Update auf v3 kommt: Connection delayed und nach ca 5 sekunden Timeout. Mehr nicht.
Web-Server-Neustart bringt nichts. Jarvis Neustart auch nicht.
Adresszeile: http://192.168.178.70:8082/jarvis/index.html#/ ?

**Reply by Zefau** [reply]
Jo, http hat ein Fehler. Behebe ich heute abend. Sonst sollte https funktionieren

**Reply by mcuiobroker** [reply]
https kommt auch wieder Connection delayed und timeout
Adresszeile: https://192.168.178.70:8082/jarvis/index.html#/

**Reply by Zefau** [reply]
mit  v1.0.0-alpha.1 ?

**Reply by mcuiobroker** [reply]
Sollte es nicht v3.0.0-alpha.1 sein?

**Reply by mcuiobroker** [reply]
Trotzdem Timeout (nach ca 10Sekunden)

**Reply by Zefau** [reply]
ach fuck stimmt; zu viel hin und her geändert

**Reply by Zefau** [reply]
Du verbindest über http? Wie greifst du auf jarvis zu? Vom ioBroker Dashboard? Wie sieht dort die URL aus, die aufgerufen wird?

**Reply by mcuiobroker** [reply]
Hatte beides versucht normalerweise über http
vom Dashboard ja
http://192.168.178.70:8082/jarvis/index.html#/
auch wenn man nur http://192.168.178.70:8082/jarvis/index.html eingibt wird daraus die obere Adresse gemacht.

**Reply by Zefau** [reply]
Was steht in der Browser Console und im ioBroker Log?

Über das ioBroker Dashboard sollte jarvis mit den notwendigen Verbindungseinstellungen aufgerufen werden.
Bei mir sieht es so aus:
![image](https://user-images.githubusercontent.com/14368811/120905488-9e46f380-c652-11eb-8f84-deeb49576b2d.png)

Wenn du allerdings kein SSL / HTTPs benutzt, sollte es trotzdem gehen.

**Reply by mcuiobroker** [reply]
Using instance 0.
​ Connecting to https://192.168.178.70:8400 ..
​ Connected to https://192.168.178.70:8400.
index.html#/:1 Uncaught SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at s.convertLayout (app.9869f3da.js:1)
    at Object.callback (app.9869f3da.js:1)
    at d.processIncomingMessage (app.9869f3da.js:1)
    at l.<anonymous> (app.9869f3da.js:1)
    at l.i.emit (vendor.fec01e23.js:43)
    at l.emit (vendor.fec01e23.js:114)
    at l.emitEvent (vendor.fec01e23.js:43)
    at l.onevent (vendor.fec01e23.js:43)
    at l.onpacket (vendor.fec01e23.js:43)

**Reply by mcuiobroker** [reply]
Ich hab aber kein 8400 Port und auch kein https, der ist 8082

![image](https://user-images.githubusercontent.com/84818934/120905778-b3bd1d00-c654-11eb-9f0f-edb24f2da258.png)

**Reply by Zefau** [reply]
Port 8400 ist von jarvis direkt für socket-io. Die Verbindung funktioniert auch, aber das JSON des Layouts ist falsch. Bist du von v2 gekommen oder von einer Neuinstallation?

**Reply by mcuiobroker** [reply]
War vorher direkt von v2 mit update und hab dann die Inhalte gelöscht, da es ja Probleme gab und hab versucht es damit in den Griff zu bekommen.

**Reply by mcuiobroker** [reply]
Hab die DP gefüllt mit Origdaten:
![image](https://user-images.githubusercontent.com/84818934/120906025-6e99ea80-c656-11eb-817a-cc7f26a6607f.png)

**Reply by Zefau** [reply]
Dann geht's jetzt? Sieht zumindest gut aus, wenn auch scheinbar mit Fehlern. Gibt es Fehler in der Browser-Konsole?

**Reply by mcuiobroker** [reply]
StateListHorizontal wird garnicht angezeigt
Alle Widgets werden in gleichgroße Fenster gepackt.
iFrame (Radar) wird nicht gefüllt und sucht ohne Ende

**Reply by mcuiobroker** [reply]
Converted DEVICES Object
​ TypeError: this.device.getState is not a function
    at s.created (app.9869f3da.js:1)
    at ie (vendor.fec01e23.js:14)
    at Pn (vendor.fec01e23.js:14)
    at s.Ti.t._init (vendor.fec01e23.js:14)
    at new s (vendor.fec01e23.js:14)
    at an (vendor.fec01e23.js:14)
    at init (vendor.fec01e23.js:14)
    at p (vendor.fec01e23.js:14)
    at f (vendor.fec01e23.js:14)
    at E (vendor.fec01e23.js:14)
oe @ vendor.fec01e23.js:14
index.html#/:1 Failed to decode downloaded font: http://192.168.178.70:8082/jarvis/fonts/flUhRq6tzZclQEJ-Vdg-IuiaDsNcIhQ8tQ.c934fe77.woff2
index.html#/:1 OTS parsing error: Size of decompressed WOFF 2.0 is less than compressed size
index.html#/:1 Failed to decode downloaded font: http://192.168.178.70:8082/jarvis/fonts/materialdesignicons-webfont.e9db4005.woff2
index.html#/:1 OTS parsing error: Size of decompressed WOFF 2.0 is less than compressed size
index.html#/:1 Failed to decode downloaded font: http://192.168.178.70:8082/jarvis/fonts/materialdesignicons-webfont.e9db4005.woff2
index.html#/:1 OTS parsing error: Size of decompressed WOFF 2.0 is less than compressed size
index.html#/:1 Failed to decode downloaded font: http://192.168.178.70:8082/jarvis/fonts/flUhRq6tzZclQEJ-Vdg-IuiaDsNcIhQ8tQ.c934fe77.woff2
index.html#/:1 OTS parsing error: Size of decompressed WOFF 2.0 is less than compressed size

**Reply by mcuiobroker** [reply]
Bei allen customHTML kommt:
![image](https://user-images.githubusercontent.com/84818934/120906134-3c3cbd00-c657-11eb-8e7b-7dfb5cd585dc.png)

**Reply by mcuiobroker** [reply]
Icon-Ladezeit ist sehr lang (10 Sekunden), teilweise werden sie nicht gefunden: Hängt glaub ich mit den nicht gefundenen Objekten (Origdaten auf Testsystem) zusammen.
![image](https://user-images.githubusercontent.com/84818934/120906166-75752d00-c657-11eb-9fa9-b2943701bf8a.png)
Er holt sich die Icons anscheinend auch immer wieder neu. Sobald man auf einen anderen TAB wechselt und wieder zurück geht, holt sich die Icons trortzdem neu.

**Reply by Zefau** [reply]
> StateListHorizontal wird garnicht angezeigt

Das StateListHorizontal Modul gibt es nicht mehr. Dafür gibt es nun HomeKitTile.

>  Alle Widgets werden in gleichgroße Fenster gepackt.

Das ist nach dem Konvertieren so. Muss dann danach individualisiert werden.

>  iFrame (Radar) wird nicht gefüllt und sucht ohne Ende

Das und all die anderen Fehler liegt an `TypeError: this.device.getState is not a function`. Irgendwie wurden deine Geräte scheinbar nicht korrekt konvertiert. Wie sieht denn bei dir das Devices JSON aus? Steht da `version: 3` drin?

**Reply by Zefau** [reply]
ah, ich glaube es liegt daran, dass bei dir `jarvis.0.widgets` nicht angelegt wurde?

**Reply by mcuiobroker** [reply]
![image](https://user-images.githubusercontent.com/84818934/120906295-82465080-c658-11eb-8a05-9e16cd82c5d3.png)

**Reply by mcuiobroker** [reply]
![image](https://user-images.githubusercontent.com/84818934/120906299-90946c80-c658-11eb-9bde-7767144858db.png)

**Reply by mcuiobroker** [reply]
jarvis.0.widgets wird auch nicht angelegt, wenn man in die Konfiguration einmal reingeht und speichert.

**Reply by Zefau** [reply]
Zieh dir mal die alpha.3 und starte nochmal mit der v2 Konfiguration. jarvis konvertiert beim Start dann in die v3 und befüllt `jarvis.0.widgets`.

**Reply by mcuiobroker** [reply]
widgets werden jetzt angelegt. iFrame sucht trotzdem weiter, keine Anzeige.

**Reply by mcuiobroker** [reply]
Kurzer Eindruck:
![jarvis3_iFrameSucht](https://user-images.githubusercontent.com/84818934/120906623-4b256e80-c65b-11eb-9b81-957d8182b7b4.gif)

**Reply by mcuiobroker** [reply]
Gibt es eine Möglichkeit dem Testsystem die Objekte vom Prodsystem mit einem Befehl zugänglich zu machen?
Testsystem : 192.168.178.70
Prodsystem: 192.168.178.89

**Reply by mcuiobroker** [reply]
Im Widget iFrame kann ich keine URL mehr eingeben, wie muss man diese jetzt angeben? Auch im Gerät kann ich keine URL angeben. Sonst war es möglich:
![image](https://user-images.githubusercontent.com/84818934/120907140-77db8500-c65f-11eb-92e7-408d16851eb0.png)

![image](https://user-images.githubusercontent.com/84818934/120907033-7e1d3180-c65e-11eb-9357-fe680b39a284.png)

Wo soll man im Gerät die URL eingeben?
![image](https://user-images.githubusercontent.com/84818934/120907090-161b1b00-c65f-11eb-9ed9-483b0467e7fb.png)



Bei den Geräten fehlen die Icons (Gewerke):
![image](https://user-images.githubusercontent.com/84818934/120907059-c63c5400-c65e-11eb-8b1c-fe79c81ed81d.png)


Beim Datenpunkt Explorer falsche Sortierung:
![image](https://user-images.githubusercontent.com/84818934/120907073-fb48a680-c65e-11eb-8578-e5fa929bfc44.png)

**Reply by mcuiobroker** [reply]
Adapterstatus:
![image](https://user-images.githubusercontent.com/84818934/120907251-4e6f2900-c660-11eb-81d5-ea7e2eed7133.png)
Man muss dort die Adapter eintragen. In v2.20 musste man das nicht, es wurden dann alle angezeigt. 
![image](https://user-images.githubusercontent.com/84818934/120907273-73fc3280-c660-11eb-8a1c-1f4c171eda88.png)

![image](https://user-images.githubusercontent.com/84818934/120907278-81b1b800-c660-11eb-8530-a5113b293367.png)
Hier nur wenn ich sie eintrage:
![image](https://user-images.githubusercontent.com/84818934/120907293-98f0a580-c660-11eb-9970-03798dabf006.png)

**Reply by Zefau** [reply]
Danke für die Hinweise. Die Bugs kannst du (zukünftig) im Issue reporten. Am besten als "Task-List" (mit dieser Checkbox), dann hake ich ab :-)

Die genannten Sachen habe ich mit alpha.4 behoben.

Grundsätzlich sieht es bei dir aber so aus, als würden die Geräte nicht geladen werden, daher auch die Jahresangaben mit 51 Jahren und das z.T. unendliche Laden. Hast du noch Fehler in der Browser-Konsole mit alpha-4?

**Reply by mcuiobroker** [reply]
Fehler auf einer Seite mit vielen customHTML:
Using instance 0.
vendor.da99f71a.js:144 Connecting to http://192.168.178.70:8400 ..
vendor.da99f71a.js:144 Connected to http://192.168.178.70:8400.
6Failed to decode downloaded font: <URL>
6OTS parsing error: Size of decompressed WOFF 2.0 is less than compressed size
2vendor.da99f71a.js:144 TypeError: Cannot read property 'toString' of undefined
    at s.inputStyle (app.9073362f.js:1)
    at ii.get (vendor.da99f71a.js:14)
    at ii.evaluate (vendor.da99f71a.js:14)
    at s.inputStyle (vendor.da99f71a.js:14)
    at s.i (app.9073362f.js:1)
    at vendor.da99f71a.js:96
    at Jt (vendor.da99f71a.js:96)
    at s.e.render (vendor.da99f71a.js:96)
    at s.Tn.t._render (vendor.da99f71a.js:14)
    at s.i (vendor.da99f71a.js:14)
    at ii.get (vendor.da99f71a.js:14)
    at new ii (vendor.da99f71a.js:14)
    at Fn (vendor.da99f71a.js:14)
    at s.Ii.$mount (vendor.da99f71a.js:14)
    at init (vendor.da99f71a.js:14)
    at f (vendor.da99f71a.js:14)
    at p (vendor.da99f71a.js:14)
    at b (vendor.da99f71a.js:14)
    at p (vendor.da99f71a.js:14)
    at b (vendor.da99f71a.js:14)
    at p (vendor.da99f71a.js:14)
    at s.__patch__ (vendor.da99f71a.js:14)
    at s.t._update (vendor.da99f71a.js:14)
    at s.i (vendor.da99f71a.js:14)
    at ii.get (vendor.da99f71a.js:14)
    at new ii (vendor.da99f71a.js:14)
    at Fn (vendor.da99f71a.js:14)
    at s.Ii.$mount (vendor.da99f71a.js:14)
    at init (vendor.da99f71a.js:14)
    at f (vendor.da99f71a.js:14)
    at p (vendor.da99f71a.js:14)
    at b (vendor.da99f71a.js:14)
    at p (vendor.da99f71a.js:14)
    at s.__patch__ (vendor.da99f71a.js:14)
    at s.t._update (vendor.da99f71a.js:14)
    at s.i (vendor.da99f71a.js:14)
    at ii.get (vendor.da99f71a.js:14)
    at new ii (vendor.da99f71a.js:14)
    at Fn (vendor.da99f71a.js:14)
    at s.Ii.$mount (vendor.da99f71a.js:14)
    at init (vendor.da99f71a.js:14)
    at f (vendor.da99f71a.js:14)
    at p (vendor.da99f71a.js:14)
    at s.__patch__ (vendor.da99f71a.js:14)
    at s.t._update (vendor.da99f71a.js:14)
    at s.i (vendor.da99f71a.js:14)
    at ii.get (vendor.da99f71a.js:14)
    at new ii (vendor.da99f71a.js:14)
    at Fn (vendor.da99f71a.js:14)
    at s.Ii.$mount (vendor.da99f71a.js:14)
(anonymous) @ vendor.da99f71a.js:144
4vendor.da99f71a.js:144 TypeError: Cannot read property 'toString' of undefined
    at s.inputStyle (app.9073362f.js:1)
    at ii.get (vendor.da99f71a.js:14)
    at ii.evaluate (vendor.da99f71a.js:14)
    at s.inputStyle (vendor.da99f71a.js:14)
    at s.i (app.9073362f.js:1)
    at vendor.da99f71a.js:96
    at Jt (vendor.da99f71a.js:96)
    at s.e.render (vendor.da99f71a.js:96)
    at s.Tn.t._render (vendor.da99f71a.js:14)
    at s.i (vendor.da99f71a.js:14)
    at ii.get (vendor.da99f71a.js:14)
    at ii.run (vendor.da99f71a.js:14)
    at qn (vendor.da99f71a.js:14)
    at Array.<anonymous> (vendor.da99f71a.js:14)
    at ue (vendor.da99f71a.js:14)
(anonymous) @ vendor.da99f71a.js:144
vendor.da99f71a.js:144 UPDATE 680fdb12-afa4-49d2-aed6-050252ba3160 OS_STADT_cases7_per_100K
(anonymous) @ vendor.da99f71a.js:144

**Comment by WW1983** [comment]
Bei mir zeigt er den gleichen Fehler. Auch mit https funktioniert es nicht

**Comment by mcuiobroker** [comment]
Kannst du bitte eine kurze Darstellung oder Überblick geben, was sich geändert haben soll, was wude verbessert usw. Damit man auch schon mal auf die Zeiten oder sonstiges achtet. Evtl in Abhängigkeit von Widgets?

**Comment by WW1983** [comment]
Ich habe es mal getestet und mir sind auf Anhieb folgende Fehler aufgefallen:

- Wenn ich im Menü eine neue Seite anlege und auf diese gehe, dann gibt es keine Möglichkeit zurück zu kommen

- in der oberen Menüleiste kann man sich nicht nur die Icons anzeigen lassen. Die Einstellung funktioniert leider nicht 

- wenn ich in der Menüleiste viele Reiter habe, dann wird ganz rechts "more" angezeigt. Dieser Button funktioniert nicht zuverlässig. Teilweise muss ich 3 mal tippen (auf dem iPad) damit es aufgeht 

- ich kann JARVIS nicht auf dem Bildschirm als "App" (auf Apple Geräten) ablegen 

- die Ansicht auf dem iPhone ist nicht zu gebrauchen. Leider finde ich keine Möglichkeit es anders einzustellen ![image](https://user-images.githubusercontent.com/56581564/120917307-d7ad4c80-c6ae-11eb-9e5c-0efde37c7fe8.png)

**Comment by RkcCorian** [comment]
Hallo Zusammen,
aktuell läuft bei mir v2.1.0 ohne Probleme. Für eine spätere Migration auf v3.x... wäre es besser von v2.2.1 (welche inzw. zum Update angeboten wird) zu Aktualisieren um später einfach auf v3.x zu kommen oder macht das keinen Unterschied und ich soll erstmal auf v2.1.0 bleiben?
Vielen Dank im Voraus!
